### PR TITLE
Publish Quickstart images to ghcr

### PR DESCRIPTION
**Pull request by tanvigour**
Signed-off-by: tanvigour <tanvi.gour@gmail.com>

# Description

_Please explain the changes you've made_

## Issue reference

We strive to have all PRs being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #567

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] The quickstart code compiles correctly
* [ ] You've tested new builds of the quickstart if you changed quickstart code
* [ ] You've updated the quickstart's README if necessary
* [ ] If you have changed the steps for a quickstart be sure that you have updated the automated validation accordingly. All of our quickstarts have annotations that allow them to be executed automatically as code. For more information see [mechanical-markdown](https://github.com/dapr/mechanical-markdown#readme). For user guide with examples see [Examples](https://github.com/dapr/mechanical-markdown/tree/main/examples#mechanical-markdown-by-example).

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -55,57 +62,146 @@ jobs:
...
-      - name: publish image to dockerhub
+     - name: GitHub container registry login
+        if: matrix.target_os != 'darwin'
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build and push docker images to GitHub container registry
+        if: matrix.target_os != 'darwin'
```
Comment (artursouza) [inline .github/workflows/build.yml:79]: Tanvi, this change does not make sense because it is cloning all the workflow from dapr/dapr. We need to change the current quickstart workflow to do a push to GHCR in addition to Docker Hub. The change is probably much smaller.

**Review by artursouza — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -55,57 +62,146 @@ jobs:
...
-          SAMPLE_LIST=(hello-kubernetes distributed-calculator pub-sub bindings observability secretstore)
-
-          for sample in "${SAMPLE_LIST[@]}"; do
-            echo "Push image for ${sample}..."
-            pushd ${sample}
-            make push
-            popd
-          done
+          echo "Build docker image and push image..."
+          make docker-push TARGET_OS=${{ matrix.target_os }} TARGET_ARCH=${{ matrix.target_arch }} DAPR_REGISTRY=ghcr.io/${{ github.repository_owner }} DAPR_TAG=${{ env.REL_VERSION }}
```
Comment (pravinpushkar) [inline .github/workflows/build.yml:82]: We can check this file for what make command we already have for pushing images.
https://github.com/dapr/quickstarts/blob/870ed621b843708cab9a78f7ec4cd122575957c2/docker.mk#L1
`make docker-push`  is not here I guess. We can think of adding ` docker-push` or reuse what we have already. 
Overall flow would look like this - 

1. Build docker images - Already being done
2. Docker login - Already being done
3. ghcr login - It is added in this PR
4. push images to docker - Already being done
5. push images to ghcr - We only need to add this.

We can combine 4 & 5 together or keep it separate.

File: .github/workflows/build.yml (hunk 1 above)
Comment (pravinpushkar) [reply .github/workflows/build.yml:79]: @artursouza As discussed earlier, In case of Docker we publish quickstarts images to a different repository than Dapr-
https://github.com/dapr/quickstarts/blob/870ed621b843708cab9a78f7ec4cd122575957c2/docker.mk#L8
With this PR, quickstarts images also will be published at the same place as Dapr components.

**Review by paulyuk — COMMENTED** [review]
@tanvigour and @artursouza the docker images I'm most worried about are the ones referenced in compose file for `kafka` and `zookeeper`.  Are we porting these two aimages to GHCR also for the sake of test stability? https://github.com/dapr/quickstarts/blob/master/bindings/docker-compose-single-kafka.yml#L8

**Comment by tanvigour** [comment]
> @tanvigour and @artursouza the docker images I'm most worried about are the ones referenced in compose file for `kafka` and `zookeeper`. Are we porting these two aimages to GHCR also for the sake of test stability? https://github.com/dapr/quickstarts/blob/master/bindings/docker-compose-single-kafka.yml#L8

@artursouza what do you think?

File: .github/workflows/build.yml (hunk 1 above)
Comment (artursouza) [reply .github/workflows/build.yml:79]: Yes, they will. We just add them under `/samples` folder. I talked to @yaron2 about that already. This PR, on the other hand, does not work at all for quickstarts.

**Comment by artursouza** [comment]
> > @tanvigour and @artursouza the docker images I'm most worried about are the ones referenced in compose file for `kafka` and `zookeeper`. Are we porting these two aimages to GHCR also for the sake of test stability? https://github.com/dapr/quickstarts/blob/master/bindings/docker-compose-single-kafka.yml#L8
> 
> @artursouza what do you think?

We can put those under the samples/3rdparty/ folder, this way we can also pull them from there.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -68,13 +67,37 @@ jobs:
...
           SAMPLE_LIST=(hello-kubernetes distributed-calculator pub-sub bindings observability secretstore)
-
           for sample in "${SAMPLE_LIST[@]}"; do
             echo "Push image for ${sample}..."
             pushd ${sample}
             make push
             popd
           done
+      - name: GitHub container registry login
+        if: matrix.target_os != 'darwin' && github.event_name != 'pull_request'
```
Comment (greenie-msft) [inline .github/workflows/build.yml:77]: Nit, but do we need this OS check?

Comment (artursouza) [reply .github/workflows/build.yml:77]: Yes. We don't want to publish the image per OS build.

File: docker.mk [hunk 4]
```diff
@@ -6,6 +6,7 @@
 # Common make targets for samples' Docker images.
 
 SAMPLE_REGISTRY        ?= docker.io/dapriosamples
+GHRC_REGISTRY          ?= ghrc.io/dapr/samples
```
Comment (artursouza) [inline docker.mk:9]: We can revert the changes in this file and simply set SAMPLE_REGISTRY=ghrc.io/dapr/samples in the workflow and reuse the existing make targets.

**Review by artursouza — CHANGES_REQUESTED** [review]

**Review by artursouza — APPROVED** [review]
